File: 60.js
function minTime(n, deps) {
  const graph = Array.from({ length: n + 1 }, () => []);
  const indeg = Array(n + 1).fill(0);
  const time = Array(n + 1).fill(1); // Each task takes 1 unit

  // Build graph
  for (const [u, v] of deps) {
    graph[u].push(v);
    indeg[v]++;
  }

  // Queue for tasks with no dependencies
  const queue = [];
  for (let i = 1; i <= n; i++) {
    if (indeg[i] === 0) queue.push(i);
  }

  // Topological Sort + DP
  while (queue.length) {
    const curr = queue.shift();
    for (const nxt of graph[curr]) {
      indeg[nxt]--;
      time[nxt] = Math.max(time[nxt], time[curr] + 1);
      if (indeg[nxt] === 0) queue.push(nxt);
    }
  }

  return Math.max(...time.slice(1));
}

// -------------------- DRIVER CODE --------------------
const deps1 = [[1, 3], [2, 3]];
console.log("Minimum Time:", minTime(3, deps1)); // Output: 2

const deps2 = [[1, 2], [1, 3], [3, 4], [2, 5], [4, 5]];
console.log("Minimum Time:", minTime(5, deps2)); // Output: 4

File: 60.py
from collections import defaultdict, deque

def minTime(n, deps):
    # Step 1: Build graph and in-degree array
    graph = defaultdict(list)
    indegree = [0] * (n + 1)
    
    for u, v in deps:
        graph[u].append(v)
        indegree[v] += 1

    # Step 2: Initialize queue with tasks having no dependencies
    q = deque()
    time = [1] * (n + 1)   # each task takes 1 unit of time
    
    for i in range(1, n + 1):
        if indegree[i] == 0:
            q.append(i)

    # Step 3: Topological sort + DP for completion time
    while q:
        curr = q.popleft()
        for nxt in graph[curr]:
            indegree[nxt] -= 1
            # Update time for the dependent task
            time[nxt] = max(time[nxt], time[curr] + 1)
            if indegree[nxt] == 0:
                q.append(nxt)

    # Step 4: Answer = maximum completion time
    return max(time[1:])

# Example 1
print(minTime(3, [[1, 3], [2, 3]]))   # Output: 2

# Example 2
print(minTime(5, [[1, 2], [1, 3], [3, 4], [2, 5], [4, 5]]))   # Output: 4
